### PR TITLE
Update prosemirror-* type definitions, add comments

### PR DESCRIPTION
**Pull request by timjb**
The type definitions have been generated using [getdocs2ts](https://github.com/bradleyayers/getdocs2ts) from the getdocs comments in the ProseMirror source.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/prosemirror-collab/index.d.ts [hunk 1]
```diff
@@ -1,14 +1,43 @@
...
 
-import { Plugin, EditorState, Transaction } from 'prosemirror-state';
+// IMPORTANT
+// This file was generated by https://github.com/bradleyayers/getdocs2ts. Please do not edit manually.
+// When you find an error in these declarations, fix the getdocs comment upstream or 'getdocs2ts', then regenerate.
+
+import { EditorState, Plugin, Transaction } from 'prosemirror-state';
 import { Step } from 'prosemirror-transform';
 
-export function collab<T>(config?: { version?: number, clientID?: number | string }): Plugin<T>;
```
Comment (bradleyayers) [inline types/prosemirror-collab/index.d.ts]: It's a shame to lose the generic type variables. I realise it's a tricky balance between end-user benefit and cost-to-maintain. Do you have any thoughts on how to incorporate these into the build process?

File: types/prosemirror-inputrules/index.d.ts [hunk 2]
```diff
@@ -1,30 +1,111 @@
...
+   * string.
+   *
+   * Or a it can be a function, which will be called with the match
+   * array produced by
+   * [`RegExp.exec`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec),
+   * as well as the start and end of the matched range, and which can
+   * return a [transaction](#state.Transaction) that describes the
+   * rule's effect, or null to indicate the input was not handled.
+   */
+  constructor(match: RegExp, handler: string | ((state: EditorState, match: string[], start: number, end: number) => Transaction | null | void));
```
Comment (bradleyayers) [inline types/prosemirror-inputrules/index.d.ts]: Do you think we should include `undefined` in these return types (in addition to void)?

File: types/prosemirror-model/index.d.ts [hunk 3]
```diff
@@ -1,259 +1,1290 @@
...
+ * Nodes are persistent data structures. Instead of changing them, you
+ * create new ones with the content you want. Old ones keep pointing
+ * at the old document shape. This is made cheaper by sharing
+ * structure between the old and new data as much as possible, which a
+ * tree shape like this (without back pointers) makes easy.
+ *
+ * **Do not** directly mutate the properties of a `Node` object. See
+ * [the guide](/docs/guide/#doc) for more information.
+ */
+declare class ProsemirrorNode {
```
Comment (bradleyayers) [inline types/prosemirror-model/index.d.ts]: Did you hit any gotchas with going this route?

**Review by bradleyayers — APPROVED** [review]
Awesome work!

**Comment by bradleyayers** [comment]
I'm not sure how the "👍 or 👎" is supposed to work, or if it's still relevant given Github has review tools now. The instructions from the bot are a big unclear to me.

File: types/prosemirror-inputrules/index.d.ts (hunk 2 above)
Comment (timjb) [reply types/prosemirror-inputrules/index.d.ts]: The `void` type already includes the value `undefined`.

File: types/prosemirror-collab/index.d.ts (hunk 1 above)
Comment (timjb) [reply types/prosemirror-collab/index.d.ts]: I think if we want to bring these generic type variables back, we should convince Marijn to add them to the getdocs type annotations.

File: types/prosemirror-model/index.d.ts (hunk 3 above)
Comment (timjb) [reply types/prosemirror-model/index.d.ts]: No.

**Comment by 5id** [comment]
Hey @RyanCavanaugh  and @andy-ms - these types don't seem to have been released, any ideas?

**Comment by andy-ms** [comment]
@5id thanks, it was a bug. Microsoft/types-publisher#395

**Comment by 5id** [comment]
@andy-ms no worries :) when should the fix be out?

File: types/prosemirror-tables/index.d.ts [hunk 4]
```diff
@@ -5,16 +5,15 @@
...
+import { Node as ProsemirrorNode, NodeSpec, Slice, ResolvedPos } from 'prosemirror-model';
 
 export interface TableNodesOptions {
   tableGroup?: string;
   cellContent: string;
   cellAttributes: { [key: string]: CellAttributes };
 }
 
-export type getFromDOM = (dom: dom.DOMElement) => any;
+export type getFromDOM = (dom: Element) => any;
```
Comment (Thinkscape) [inline types/prosemirror-tables/index.d.ts]: This is an invalid change unfortunately, which just broke our stuff 👎 
I'll try to get it reverted...

**Pull request merged**